src/LSP4/getAssetMetadata/getAssetMetadata.ts

```diff
@@ -0,0 +1,43 @@
+// types
+import { BytesLike, Provider } from 'ethers';
+import { ERC725YDataKeys } from '@lukso/lsp-smart-contracts';
+
+// utils
+import {
+    LSP4AssetMetadata,
+    decodeJsonUrl,
+    getErc725yContract,
+    isAssetMetadata,
+    validateIpfsUrl,
+} from '../..';
+
+// types
+import { ERC725Y } from '../../typechain';
+
+export async function getAssetMetadata(digitalAsset: ERC725Y): Promise<LSP4AssetMetadata>;
+export async function getAssetMetadata(
+    digitalAsset: ERC725Y | BytesLike,
+    provider?: Provider,
+): Promise<LSP4AssetMetadata> {
+    const digitalAssetContract: ERC725Y = provider
+        ? await getErc725yContract(digitalAsset, provider)
+        : await getErc725yContract(digitalAsset);
+
+    const assetMetadataDataValue = await digitalAssetContract.getData(
+        ERC725YDataKeys.LSP4.LSP4Metadata,
+    );
+
+    const { url } = decodeJsonUrl(assetMetadataDataValue);
```

Comment:
The new type is VerifiableURI and there are several cases
JSONURL, AssetURL or (bytes4,URI)
The code in erc725 is ```js
const decodeDataSourceWithHash = (value: string): URLDataWithHash => {
  if (value.slice(0, 6) === '0x0000') {
    // DEAL with VerifiableURI
    // NOTE: A JSONURL with a 0x00000000 verification method is invalid.

    /*
      0        1         2         3         4         5         6         7         8
      12345678901234567890123456789012345678901234567890123456789012345678901234567890
      0x0000 code
            6f357c6a hash fn [6]
                    0020 data len [14]
                        820464ddfac1be...[18 + data len]
                                                       [18 + data len]...696670733a2f2...[...rest]
    */
    const verificationMethodSignature = `0x${value.slice(6, 14)}`;
    // NOTE: verificationMethodSignature can be 0x00000000 if no verification method is used
    // this means that an invalid verification method should still return all data
    // and not be an error. It's up to the method calling this to figure out
    // whether an unknown verification method is an error or not.
    const verificationMethod = getVerificationMethod(
      verificationMethodSignature,
    );
    const encodedLength = `0x${value.slice(14, 18)}`; // Rest of data string after function hash
    const dataLength = hexToNumber(encodedLength, false) as number;
    const dataHash = `0x${value.slice(18, 18 + dataLength * 2)}`; // Get jsonHash 32 bytes
    const dataSource = hexToUtf8('0x' + value.slice(18 + dataLength * 2)); // Get remainder as URI

    return {
      verification: {
        method: verificationMethod?.name || verificationMethodSignature,
        data: dataHash,
      },
      url: dataSource,
    };
  }

  // @Deprecated code here:

  // Eventually we should no longer have JSONURL, AssetURL or (bytes4,URI)

  // DEAL with JSONURL

  const verificationMethodSignature = value.slice(0, 10);
  const verificationMethod = getVerificationMethod(verificationMethodSignature);
  const encodedData = value.slice(10); // Rest of data string after function hash

  try {
    // Special case where JSONURL is really (bytes4,URI) as specified
    // by the old version of LSP8TokenMetadataBaseURI
    // Catch error in case the buffor is not convertable to utf8.
    const dataSource = hexToUtf8('0x' + encodedData); // Get as URI
    if (encodedData.length < 64 || /^[a-z]{2,}:[/\S]/.test(dataSource)) {
      // If the verification data starts with a utf8 sequence that looks like https:/ or data: or ar:/ for example.
      return {
        verification: {
          method: NONE_VERIFICATION_METHOD,
          data: '0x',
        },
        url: dataSource,
      };
    }
  } catch {
    // ignore
  }

  const dataHash = '0x' + encodedData.slice(0, 64); // Get jsonHash 32 bytes
  const dataSource = hexToUtf8('0x' + encodedData.slice(64)); // Get remainder as URI

  return {
    verification: {
      method: verificationMethod?.name || verificationMethodSignature,
      data: dataHash,
    },
    url: dataSource,
  };
};
```

Reply:
NOTE that the verification method 0x00000000 (which is NONE) is not compatible with JSONURL or AssetURL, because for a valid VerifiableURI we expect it to start out with 0x0000 (using it as a flag)

Reply:
NOTE: When returning data with an unknown method it should return the method as is (i.e. the signature) to allow a third party to implement their own verification method. There are a few commits which return "unknown" as the method instead.

Reply:
NOTE: The workaround checking for less than 64 + 8 or regex is for (bytes4,URI) this is the format of the old LSP8 BaseURI which was defined as (bytes4,URI) where bytes4 was always a signature method but no verification data since it's not verified follow immediately by the UTF8 string.